src/core/graph.js

```diff
@@ -889,6 +889,71 @@ export class Graph {
     return result;
   }
 
+  /**
+   * Create a new graph, in which some nodes have been combined.
+   *
+   * combineNodes allows collapsing nodes together. For example, if we had a graph
+   * with nodes A, B and C, then we could collapse A into B, creating a new
+   * graph with just nodes B and C. The edges in the graph are re-written in concert;
+   * for example, an edge from A to B would become a loop edge from B to itself.
+   *
+   * combineNodes' motivating use cases is collapsing different user identities together.
+   * For example, if a user is represented on both GitHub and Discourse, we can create
+   * a new node that represents that identity, and collapse their GitHub and Discourse user
+   * nodes into it. (If we chose, we could also collapse the GitHub node into the Discourse
+   * node or vice-versa.)
+   *
+   * As its argument, combineNodes takes a map which maps a "src" node address
+   * (the address to be collapsed into another node) and a "dst" node address
+   * (the node to collapse into). As currently implemented, if a node is ever a
+   * dst, it can never be a src, although the code could be re-written to
+   * support this. Also, all "dst" nodes much actually be present in the graph
+   * (although this constraint could also be relaxed).
+   *
+   * The resultant Graph is distinct from the input graph, which is never
+   * modified. combineNodes runs in O(n+e+k), where `n` is the number of nodes,
+   * `e` is the number of edges, and `k` is the number of combinations. If
+   * needed, we can improve the peformance by mutating the original graph
+   * instead of creating a new one.
+   */
+  combineNodes(combinations: Map<NodeAddressT, NodeAddressT>): Graph {
```

Comment:
I’d prefer to call this `contractNodes` so that the meaning is
immediately obvious. If you object for some reason, the doc comment
should at least mention contraction for greppability.